src/Neo/SmartContract/Native/NeoToken.cs

```diff
@@ -327,13 +327,34 @@ public BigInteger UnclaimedGas(DataCache snapshot, UInt160 account, uint end)
             return CalculateBonus(snapshot, state, end);
         }
 
+        [ContractMethod(Hardfork.HF_Echidna, RequiredCallFlags = CallFlags.States | CallFlags.AllowCall | CallFlags.AllowNotify)]
+        private async ContractTask OnNEP17Payment(ApplicationEngine engine, UInt160 from, BigInteger amount, StackItem data)
+        {
+            if (engine.CallingScriptHash != GAS.Hash)
+                throw new InvalidOperationException("only GAS is accepted");
+
+            if ((long)amount != GetRegisterPrice(engine.SnapshotCache))
+                throw new ArgumentException("incorrect GAS amount for registration");
+
+            var pubkey = ECPoint.DecodePoint(data.GetSpan(), ECCurve.Secp256r1);
+
+            await GAS.Burn(engine, Hash, amount);
+
+            RegisterInternal(engine, pubkey);
```

Comment:
directly call `registerinternal` here?

Reply:
I mean, can this be mistakenly triggered? cause there is only a GAS token and amount check.

Reply:
We need to share some code between the old `registerCandidate` and new `onNEP17Payment` (two externally-available contract methods), that's `RegisterInternal` (which is private).

Reply:
i understand how you want it to work, but my concern is can we have some way to explicitly telling the system that the transaction is for registration? for instance a string "registerCandidate" in data.

Reply:
Can be done, `data` can be anything, for example an array of `["registerCandidate", pubkey]`. Here we have the minimal `data` possible (key only). It's hard to expect any extensions of this callback, but even if needed it still can be done in a compatible way (accept an array or anything else instead of key).